Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp

```diff
@@ -174,14 +174,24 @@ void BorderPainter::calculateBorderStyleColor(BorderStyle style, BoxSide side, C
 
     Operation operation = (side == BoxSide::Top || side == BoxSide::Left) == (style == BorderStyle::Inset) ? Darken : Lighten;
 
+    bool isVeryDarkColor = color.luminance() <= baseDarkColorLuminance;
+    bool isVeryLightColor = color.luminance() > baseLightColorLuminance;
+
+    // Special case very dark colors to give them extra contrast, similarly to Firefox.
```

Comment:
Seems like something like this should be written into the standard rather than being borrowed folklore like this.

I also think “similarly to Firefox” is a little bit imprecise and almost a tease. Is it exactly what Firefox is doing at some point in time? Is it different? If so, is it better or not as good? I’d like more of a clue than just “similar”.

The code below also already has a similarly vague comment with the same drawbacks. Not even particularly good grammar to say “yield a similar behavior as in FF”.

Reply:
> I also think “similarly to Firefox” is a little bit imprecise and almost a tease. Is it exactly what Firefox is doing at some point in time? Is it different? If so, is it better or not as good? I’d like more of a clue than just “similar”.

It's better than Firefox in this case, Firefox special cases the color black (rgb(0,0,0)) specifically, not colors close to black.

Reply:
> Seems like something like this should be written into the standard rather than being borrowed folklore like this.

I filed https://github.com/w3c/csswg-drafts/issues/9393 earlier today